### ISSUE TITLE
Linear gradient with repeated stops with `offset="0"` does not render

### ISSUE THREAD

**Issue by goransle** [yWorks/svg2pdf.js#151]
**Describe the bug**
Linear gradient with repeated stops with `offset="0"` does not render.

**To Reproduce**
See this [playground](http://raw.githack.com/yWorks/svg2pdf.js/master/?svg=%253Csvg%2520viewBox%253D%25220%25200%2520300%2520150%2522%2520xmlns%253D%2522http%253A%252F%252Fwww.w3.org%252F2000%252Fsvg%2522%253E%250A%253Cdefs%2520aria-hidden%253D%2522true%2522%253E%250A%2520%2520%253ClinearGradient%2520id%253D%2522gradient%2522%2520y2%253D%25221%2522%2520x2%253D%25220%2522%2520y1%253D%25220%2522%2520x1%253D%25220%2522%253E%250A%2520%2520%2520%2520%253Cstop%2520stop-opacity%253D%25221%2522%2520stop-color%253D%2522%2523FF0000%2522%2520offset%253D%25220%2522%253E%253C%252Fstop%253E%250A%2520%2520%2520%2520%253Cstop%2520stop-opacity%253D%25221%2522%2520stop-color%253D%2522%2523FFFFFF%2522%2520offset%253D%25220%2522%253E%253C%252Fstop%253E%250A%2520%2520%2520%2520%253Cstop%2520stop-opacity%253D%25221%2522%2520stop-color%253D%2522%25237cb5ec%2522%2520offset%253D%25221%2522%253E%253C%252Fstop%253E%250A%2520%2520%253C%252FlinearGradient%253E%250A%2520%2520%253C%252Fdefs%253E%250A%2520%2520%250A%2520%2520%253Crect%2520data-z-index%253D%25221%2522%2520fill%253D%2522url%28%2523gradient%29%2522%2520height%253D%2522280%2522%2520width%253D%2522100%2522%2520y%253D%25224%2522%2520x%253D%25220%2522%253E%253C%252Frect%253E%2520%2520%250A%253C%252Fsvg%253E).

**Expected behavior**
I would have expected the PDF to render the same way as the SVG. 
This is in accordance to the following SVG spec: https://www.w3.org/TR/SVG2/pservers.html#StopNotes.

**Desktop:**
 - OS: OSX Catalina (10.15.5)
 - Browser Chrome
 - Version 84.0.4147.135

**Comment by yGuy** [comment]
I can confirm that this doesn't work. I guess the important part of the spec is the following. It also hints at the workaround or possible implementation in the conversion:

> If two gradient stops have the same offset value, then the latter gradient stop controls the color value at the overlap point. In particular:
```
<stop offset="0" stop-color="white"/>
<stop offset=".2" stop-color="red"/>
<stop offset=".2" stop-color="blue"/>
<stop offset="1" stop-color="black"/>
```
> will have approximately the same effect as:
```
<stop offset="0" stop-color="white"/>
<stop offset=".1999999999" stop-color="red"/>
<stop offset=".2" stop-color="blue"/>
<stop offset="1" stop-color="black"/>
```
>which is a gradient that goes smoothly from white to red, then abruptly shifts from red to blue, and then goes smoothly from blue to black.

**Comment by yGuy** [comment]
Interestingly this seems to work with values which are not `0` and having two different colors both at `0` really doesn't make any sense. So I would suggest to also reconsider the implementation that actually produces two different colors both at offset `0` what would be the expected behavior in the case, after all?

**Comment by goransle** [comment]
@yGuy We could replace the second offset with a very low number, but I see now that there is also a difference in the gradient produced then: [playground](http://raw.githack.com/yWorks/svg2pdf.js/master/?svg=%253Csvg%2520xmlns%253D%2522http%253A%252F%252Fwww.w3.org%252F2000%252Fsvg%2522%2520viewBox%253D%25220%25200%2520300%2520150%2522%253E%250A%253Cdefs%2520aria-hidden%253D%2522true%2522%253E%250A%2520%2520%253ClinearGradient%2520x1%253D%25220%2522%2520y1%253D%25220%2522%2520x2%253D%25220%2522%2520y2%253D%25221%2522%2520id%253D%2522gradient%2522%253E%250A%2520%2520%2520%2520%253Cstop%2520offset%253D%25220%2522%2520stop-color%253D%2522%2523FF0000%2522%2520stop-opacity%253D%25221%2522%253E%253C%252Fstop%253E%250A%2520%2520%2520%2520%253Cstop%2520offset%253D%25220.0000000000000000001%2522%2520stop-color%253D%2522%2523FFFFFF%2522%2520stop-opacity%253D%25221%2522%253E%253C%252Fstop%253E%250A%2520%2520%2520%2520%253Cstop%2520offset%253D%25221%2522%2520stop-color%253D%2522%25237cb5ec%2522%2520stop-opacity%253D%25221%2522%253E%253C%252Fstop%253E%250A%2520%2520%253C%252FlinearGradient%253E%250A%2520%2520%253C%252Fdefs%253E%250A%2520%2520%250A%2520%2520%253Crect%2520x%253D%25220%2522%2520y%253D%25224%2522%2520width%253D%2522100%2522%2520height%253D%2522280%2522%2520fill%253D%2522url%28%2523gradient%29%2522%2520data-z-index%253D%25221%2522%253E%253C%252Frect%253E%2520%2520%250A%253C%252Fsvg%253E).

**Comment by yGuy** [comment]
You're right; both (at least Chrome) and the PDF are wrong and behave against the spec. Even with both offsets at exactly 0 you should not be seeing this thin red line in the SVG and the PDF seems to round up the value to about 0.05
But at least with this workaround the PDF isn't completely broken.

**Comment by HackbrettXXX** [comment]
jsPDF linearly samples the gradient colors, which is why very hard edges are washed out. As a fix we should probably implement the workaround @yGuy mentioned. When there are multiple stops at `0` or `1`, we can ignore the "outer" stops.
